posthog/hogql/printer.py

```diff
@@ -1,196 +1,298 @@
         if node.op == ast.BinaryOperationType.Add:
-            response = f"plus({print_ast(node.left, stack, context, dialect)}, {print_ast(node.right, stack, context, dialect)})"
+            return f"plus({self.visit(node.left)}, {self.visit(node.right)})"
         elif node.op == ast.BinaryOperationType.Sub:
-            response = f"minus({print_ast(node.left, stack, context, dialect)}, {print_ast(node.right, stack, context, dialect)})"
+            return f"minus({self.visit(node.left)}, {self.visit(node.right)})"
         elif node.op == ast.BinaryOperationType.Mult:
-            response = f"multiply({print_ast(node.left, stack, context, dialect)}, {print_ast(node.right, stack, context, dialect)})"
+            return f"multiply({self.visit(node.left)}, {self.visit(node.right)})"
         elif node.op == ast.BinaryOperationType.Div:
-            response = f"divide({print_ast(node.left, stack, context, dialect)}, {print_ast(node.right, stack, context, dialect)})"
+            return f"divide({self.visit(node.left)}, {self.visit(node.right)})"
         elif node.op == ast.BinaryOperationType.Mod:
-            response = f"modulo({print_ast(node.left, stack, context, dialect)}, {print_ast(node.right, stack, context, dialect)})"
+            return f"modulo({self.visit(node.left)}, {self.visit(node.right)})"
         else:
             raise ValueError(f"Unknown BinaryOperationType {node.op}")
-    elif isinstance(node, ast.And):
-        response = f"and({', '.join([print_ast(operand, stack, context, dialect) for operand in node.exprs])})"
-    elif isinstance(node, ast.Or):
-        response = f"or({', '.join([print_ast(operand, stack, context, dialect) for operand in node.exprs])})"
-    elif isinstance(node, ast.Not):
-        response = f"not({print_ast(node.expr, stack, context, dialect)})"
-    elif isinstance(node, ast.OrderExpr):
-        response = f"{print_ast(node.expr, stack, context, dialect)} {node.order}"
-    elif isinstance(node, ast.CompareOperation):
-        left = print_ast(node.left, stack, context, dialect)
-        right = print_ast(node.right, stack, context, dialect)
+
+    def visit_and(self, node: ast.And):
+        return f"and({', '.join([self.visit(operand) for operand in node.exprs])})"
+
+    def visit_or(self, node: ast.Or):
+        return f"or({', '.join([self.visit(operand) for operand in node.exprs])})"
+
+    def visit_not(self, node: ast.Not):
+        return f"not({self.visit(node.expr)})"
+
+    def visit_order_expr(self, node: ast.OrderExpr):
+        return f"{self.visit(node.expr)} {node.order}"
+
+    def visit_compare_operation(self, node: ast.CompareOperation):
+        left = self.visit(node.left)
+        right = self.visit(node.right)
         if node.op == ast.CompareOperationType.Eq:
             if isinstance(node.right, ast.Constant) and node.right.value is None:
-                response = f"isNull({left})"
+                return f"isNull({left})"
```

Comment:
ooo tricky, special handling for nulls because the user did something that probably wouldn't work?

iirc it's not invalid Clickhouse SQL to have `equals(NULL, 1)`

Reply:
Hmm... I have nothing against splitting this back to a separate "IsNull" comparison. This basically came out of a property filter. The logic is that to check if anything is or isn't null, you can't just `== null` or `!= null` you always have to change it to `is` or `is not`. This effectively makes comparing with null work.

But, that might not be what's expected, so let's revisit. I'll add it to the list of checkboxes.